language/solana/move-to-solana/src/stackless/extensions.rs

```diff
@@ -21,19 +21,101 @@ pub impl<'a> ModuleEnvExt for mm::ModuleEnv<'a> {
 #[extension_trait]
 pub impl<'a> FunctionEnvExt for mm::FunctionEnv<'a> {
     fn llvm_symbol_name(&self, tyvec: &[mty::Type]) -> String {
-        let mut name = self.get_full_name_str();
+        let name = self.get_full_name_str();
         if name == "<SELF>::<SELF>" {
             // fixme move-model names script fns "<SELF>".
             // we might want to preserve the actual names
             "main".to_string()
         } else {
-            for ty in tyvec {
-                name += &format!("_{}", ty.display(&self.get_type_display_ctx()))
-            }
-            name.replace([':', '<', '>'], "_").replace(", ", "_")
+            self.llvm_symbol_name_full(tyvec)
         }
     }
 
+    /// Generate a symbol name that is less than 64 bytes long.
+    ///
+    /// The rbpf VM supports symbol names up to 64 bytes in length
+    /// (defined by SYMBOL_NAME_LENGTH_MAXIMUM within rbpf),
+    /// the last byte of which is 0, so we have 63 bytes to work with.
+    /// We also need to ensure that names are unique with no collisions
+    /// across modules.
+    ///
+    /// The scheme is:
+    ///
+    /// - 16 bytes - The low 8 bytes of the module address, hex encoded.
```

Comment:
Yes. All the human-readable parts of the symbol encoding are lossy and just for human readability. The hash is the only thing in the encoding that ensures uniqueness.

It isn't possible to hex-encode the full module address into symbol names because it would take more than the 63 bytes available to the symbol.